boa/src/builtins/number/tests.rs

```diff
@@ -177,12 +178,255 @@ fn to_string() {
     let float_string = forward(&mut engine, "float_string");
     let exp_string = forward(&mut engine, "exp_string");
     let neg_string = forward(&mut engine, "neg_string");
+    let nan_string = forward(&mut engine, "nan_string");
 
     assert_eq!(default_string, String::from("0"));
     assert_eq!(int_string, String::from("123"));
     assert_eq!(float_string, String::from("1.234"));
     assert_eq!(exp_string, String::from("12000"));
     assert_eq!(neg_string, String::from("-1.2"));
+    assert_eq!(nan_string, String::from("NaN"));
+}
+
+#[test]
+fn to_string_with_radix() {
+    let realm = Realm::create();
+    let mut engine = Executor::new(realm);
+    let init = r#"
+        var zero_with_radix = Number(0).toString(16);
+        var num_smaller_than_radix = Number(9).toString(16);
+        var num_with_radix = Number(90).toString(16);
+        var fraction_with_radix = Number(90.90).toString(16);
+
+
+    "#;
+
+    eprintln!("{}", forward(&mut engine, init));
+    let zero_with_radix = forward(&mut engine, "zero_with_radix");
+    let num_smaller_than_radix = forward(&mut engine, "num_smaller_than_radix");
+    let num_with_radix = forward(&mut engine, "num_with_radix");
+    let fraction_with_radix = forward(&mut engine, "fraction_with_radix");
+
+    assert_eq!(zero_with_radix, String::from("0"));
+    assert_eq!(num_smaller_than_radix, String::from("9"));
+    assert_eq!(num_with_radix, String::from("5a"));
+    assert_eq!(fraction_with_radix, String::from("5a.e66666666668"));
+
+    assert_eq!(
+        String::from("NaN"),
+        forward(&mut engine, "Number(NaN).toString()")
+    );
+    assert_eq!(
+        String::from("Infinity"),
+        forward(&mut engine, "Number(1/0).toString()")
+    );
+    assert_eq!(
+        String::from("-Infinity"),
+        forward(&mut engine, "Number(-1/0).toString()")
+    );
+    assert_eq!(
+        String::from("0"),
+        forward(&mut engine, "Number(0).toString()")
+    );
+    assert_eq!(
+        String::from("9"),
+        forward(&mut engine, "Number(9).toString()")
+    );
+    assert_eq!(
+        String::from("90"),
+        forward(&mut engine, "Number(90).toString()")
+    );
+    assert_eq!(
+        String::from("90.12"),
+        forward(&mut engine, "Number(90.12).toString()")
+    );
+    assert_eq!(
+        String::from("0.1"),
+        forward(&mut engine, "Number(0.1).toString()")
+    );
+    assert_eq!(
+        String::from("0.01"),
+        forward(&mut engine, "Number(0.01).toString()")
+    );
+    assert_eq!(
+        String::from("0.0123"),
+        forward(&mut engine, "Number(0.0123).toString()")
+    );
+    // assert_eq!(String::from("111111111111111110000"), forward(&mut engine, "Number(111111111111111111111).toString()"));
+    // assert_eq!(String::from("1.1111111111111111e+21"), forward(&mut engine, "Number(1111111111111111111111).toString()"));
```

Comment:
This tests fails i think because rusts default formatting does not match the js spec.

Reply:
> This tests fails i think because rusts default formatting does not match the js spec.

Yes. The rust formatter is not JavaScript spec compliant. Even using `format!("{:e}", num)` isn't sufficient.
We probably need too write our own f64 parser, like we did the one with bases. Here is V8 [DoubleToCString](https://chromium.googlesource.com/v8/v8/+/refs/heads/master/src/numbers/conversions.cc#938)